Organize CLI flags into groups (for better --help usability)

TheCleric opened the pull request:
This PR will improve our CLI flags in the `--help` and in the documentation. To accomplish this is was necessary to migrate from commander to yargs for CLI processing.

- CLI flags are now grouped together for easier discovery
- CLI flags are now alphabetical to help finding a desired option easier
- CLI flags are now UNDER UNIT TEST 🎉 

Here is the new output of `--help` (which is adaptive to terminal width)
```
$ nativefier --help
nativefier <targetUrl> [outputDirectory] [other options]
or
nativefier --upgrade <pathToExistingApp> [other options]

Positionals:
  targetUrl        The URL that you wish to to turn into a native app. Required if not using --upgrade  [string]
  outputDirectory  The directory to output the generated app to            [string] [default: current directory]

App Creation Options
  -a, --arch              The cpu architecture to build for
        [string] [choices: "ia32", "x64", "armv7l", "arm64"] [default: current node's compiled cpu architecture]
  -c, --conceal           packages the app source code into an asar archive           [boolean] [default: false]
  -e, --electron-version  electron version to package, without the 'v', see
                          https://github.com/electron/electron/releases                        [default: 12.0.6]
      --global-shortcuts  JSON file defining global shortcuts. See
                          https://github.com/nativefier/nativefier/blob/master/docs/api.md#global-shortcuts
                                                                                                        [string]
  -i, --icon              the icon file to use as the icon for the app (should be a .png, on macOS can also be
                          an .icns)                                                                     [string]
  -n, --name              Name to use for the app [string] [default: The title of the page passed via targetUrl]
      --no-overwrite      do not overwrite output directory if it already exists      [boolean] [default: false]
  -p, --platform          The operating system platform to build for
       [string] [choices: "darwin", "linux", "mac", "mas", "osx", "windows"] [default: current operating system]
      --portable          Make the app store its user data in the app folder. WARNING: see
                          https://github.com/nativefier/nativefier/blob/master/docs/api.md#portable for security
                          risks                                                       [boolean] [default: false]
      --upgrade           Upgrade an app built by an older version of Nativefier.
                          You must pass the full path to the existing app executable (app will be overwritten
                          with upgraded version by default)                                             [string]
      --widevine          use a Widevine-enabled version of Electron for DRM playback (use at your own risk,
                          it's unofficial, provided by CastLabs)                      [boolean] [default: false]

App Window Options
      --always-on-top          enable always on top window                            [boolean] [default: false]
      --background-color       sets the app background color, for better integration while the app is loading.
                               Example value: '#2e2c29'                                                 [string]
      --bookmarks-menu         Path to JSON configuration file for the bookmarks menu                   [string]
      --browserwindow-options  a JSON string that will be sent directly into electron BrowserWindow options. See
                               https://github.com/nativefier/nativefier/blob/master/docs/api.md#browserwindow-op
                               tions                                                                    [string]
      --disable-context-menu   disable the context menu (right click)                 [boolean] [default: false]
      --disable-dev-tools      disable developer tools (Ctrl+Shift+I / F12)           [boolean] [default: false]
      --full-screen            always start the app full screen                       [boolean] [default: false]
      --height                 set window default height in pixels                       [number] [default: 800]
      --hide-window-frame      disable window frame and controls                      [boolean] [default: false]
  -m, --show-menu-bar          set menu bar visible                                   [boolean] [default: false]
      --max-height             set window maximum height; in pixels                [number] [default: unlimited]
      --max-width              set window maximum width in pixels                  [number] [default: unlimited]
      --maximize               always start the app maximized                         [boolean] [default: false]
      --min-height             set window minimum height in pixels                         [number] [default: 0]
      --min-width              set window minimum width in pixels                          [number] [default: 0]
      --process-envs           a JSON string of key/value pairs to be set as environment variables before any
                               browser windows are opened                                               [string]
      --single-instance        allow only a single instance of the application        [boolean] [default: false]
      --tray                   Allow app to stay in system tray. If 'start-in-tray' is set as argument, don't
                               show main window on first start                        [boolean] [default: false]
      --width                  app window default width in pixels                       [number] [default: 1280]
  -x                           set window x location in pixels from left                                [number]
  -y                           set window y location in pixels from top                                 [number]
      --zoom                   default zoom factor to use when the app is opened           [number] [default: 1]

Internal Browser Options
      --file-download-options  a JSON string of key/value pairs to be set as file download options. See
                               https://github.com/sindresorhus/electron-dl for available options.       [string]
      --honest                 prevent the normal changing of the user agent string to appear as a regular
                               Chrome browser                                         [boolean] [default: false]
      --inject                 path to a CSS/JS file to be injected. Pass multiple times to inject multiple
                               files.                                                                    [array]
      --lang                   set the language or locale to render the web site as (e.g., "fr", "en-US", "es",
                               etc.)                       [string] [default: os language at runtime of the app]
  -u, --user-agent             set the app user agent string                                            [string]

URL Handling Options
      --block-external-urls  forbid navigation to URLs not considered "internal" (see '--internal-urls').
                             Instead of opening in an external browser, attempts to navigate to external URLs
                             will be blocked                                          [boolean] [default: false]
      --internal-urls        regex of URLs to consider "internal"; all other URLs will be opened in an external
                             browser.                      [string] [default: URLs sharing the same base domain]
      --proxy-rules          proxy rules; see https://www.electronjs.org/docs/api/session#sessetproxyconfig
                                                                                                        [string]

Auth Options
      --basic-auth-password  basic http(s) auth password                                                [string]
      --basic-auth-username  basic http(s) auth username                                                [string]

Graphics Options
      --disable-gpu           disable hardware acceleration                           [boolean] [default: false]
      --enable-es3-apis       force activation of WebGL 2.0                           [boolean] [default: false]
      --ignore-gpu-blacklist  force WebGL apps to work on unsupported GPUs            [boolean] [default: false]

Caching Options
      --clear-cache      prevent the application from preserving cache between launches
                                                                                      [boolean] [default: false]
      --disk-cache-size  forces the maximum disk space (in bytes) to be used by the disk cache
                                                                            [number] [default: chromium default]

(In)Security Options
      --disable-old-build-warning-yesiknowitisinsecure  Disables warning when opening an app made with an old
                                                        version of Nativefier. Nativefier uses the Chrome
                                                        browser (through Electron), and it is dangerous to keep
                                                        using an old version of it.)  [boolean] [default: false]
      --ignore-certificate                              ignore certificate-related errors
                                                                                      [boolean] [default: false]
      --insecure                                        enable loading of insecure content
                                                                                      [boolean] [default: false]

Flash Options (DEPRECATED)
      --flash       enables Adobe Flash                                  [deprecated] [boolean] [default: false]
      --flash-path  path to Chrome flash plugin; find it in `chrome://plugins`             [deprecated] [string]

Platform Specific Options
      --app-copyright             (macOS, windows only) a human-readable copyright line for the app. Maps to
                                  `LegalCopyright` metadata property on Windows, and `NSHumanReadableCopyright`
                                  on macOS                                                              [string]
      --app-version               (macOS, windows only) the version of the app. Maps to the `ProductVersion`
                                  metadata property on Windows, and `CFBundleShortVersionString` on macOS.
                                                                                                        [string]
      --bounce                    (macOS only) make the dock icon bounce when the counter increases
                                                                                      [boolean] [default: false]
      --build-version             (macOS, windows only) The build version of the app. Maps to `FileVersion`
                                  metadata property on Windows, and `CFBundleVersion` on macOS          [string]
      --counter                   (macOS only) set a dock count badge, determined by looking for a number in the
                                  window title                                        [boolean] [default: false]
      --darwin-dark-mode-support  (macOS only) enable Dark Mode support on macOS 10.14+
                                                                                      [boolean] [default: false]
  -f, --fast-quit                 (macOS only) quit app on window close               [boolean] [default: false]
      --title-bar-style           (macOS only) set title bar style. Consider injecting custom CSS (via --inject)
                                  for better integration             [string] [choices: "hidden", "hiddenInset"]
      --win32metadata             (windows only) a JSON string of key/value pairs (ProductName, InternalName,
                                  FileDescription) to embed as executable metadata                      [string]

Debug Options
      --crash-reporter  remote server URL to send crash reports                                         [string]
      --verbose         enable verbose/debug/troubleshooting logs                     [boolean] [default: false]

Other Options
      --version  Show version number                                                                   [boolean]
      --help     Show help                                                                             [boolean]

Examples:
  nativefier <targetUrl> -n <name>                Make an app from <targetUrl> and set the application name to
                                                  <name>
  nativefier --upgrade <pathToExistingApp>        Upgrade (in place) the existing Nativefier app at
                                                  <pathToExistingApp>
  nativefier <targetUrl> -p <platform> -a <arch>  Make an app from <targetUrl> for the os <platform> and cpu
                                                  architecture <arch>
```

TheCleric commented:
@ronjouch Looks like we have a problem: `yargs supports a minimum Node.js version of 12.`

Evidently Node 10 was announced as End of Life on 4/30: https://github.com/nodejs/node/blob/master/CHANGELOG.md

Do we want to go ahead and deprecate it here as well?

TheCleric commented:
> @ronjouch Looks like we have a problem: `yargs supports a minimum Node.js version of 12.`
> 
> Evidently Node 10 was announced as End of Life on 4/30: https://github.com/nodejs/node/blob/master/CHANGELOG.md
> 
> Do we want to go ahead and deprecate it here as well?

If we do want to do so, I made a separate PR for it here: https://github.com/nativefier/nativefier/pull/1192

ronjouch commented:
> > @ronjouch Looks like we have a problem: `yargs supports a minimum Node.js version of 12.`
> > Evidently Node 10 was announced as End of Life on 4/30: https://github.com/nodejs/node/blob/master/CHANGELOG.md
> > Do we want to go ahead and deprecate it here as well?
> 
> If we do want to do so, I made a separate PR for it here: #1192



> > @ronjouch Looks like we have a problem: `yargs supports a minimum Node.js version of 12.`
> > Evidently Node 10 was announced as End of Life on 4/30: https://github.com/nodejs/node/blob/master/CHANGELOG.md
> > Do we want to go ahead and deprecate it here as well?
> 
> If we do want to do so, I made a separate PR for it here: #1192

@TheCleric I usually bump the minimum node version when it lands in the repos of the stablest laggiest conservativest OS we can expect: Debian stable. And Debian stable (currently `buster`) ships [10.24.0](https://packages.debian.org/search?suite=all&keywords=nodejs), which is still maintained for [security patches](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V10.md). That being said,

- Our users running Debian stable (if we have more than 3.4 😄) have ways to access Node 12.
- A few of our dependencies are now un-bumpable because like `yargs` they require Node 12.

, so yeah let's do this.

That being said, please don't merge https://github.com/nativefier/nativefier/pull/1192 immediately,

1. I'll cut a final release supporting 10, announcing that the next release will no longer support it
2. There are a couple of extra things to add to #1192. I'll push to your branch and, as I do so, I'll document all the knobs that needs adjusting, so that next time we bump, what needs adjustment is obvious.

ronjouch commented:
@TheCleric letting you rebase this to get a working build with node12! I'll review this tomorrow or on Monday.

TheCleric commented:
> @TheCleric letting you rebase this to get a working build with node12! I'll review this tomorrow or on Monday.

Ready whenever you are.

ronjouch commented on src/cli.test.ts:
```diff
@@ -0,0 +1,292 @@
+    initArgs(['https://www.google.com', '--help']);
+    expect(mockExit).toHaveBeenCalledTimes(1);
+    expect(mockLog).toBeCalled();
+    mockLog.mockRestore();
```
Here and below: why not move this other restore to `afterEach` ?

TheCleric replied:
> Here and below: why not move this other restore to `afterEach` ?

In this case, we created the mock for this test only. I didn't want to do it in all tests as that could hide potentially useful output if the test begins failing.

ronjouch replied:
Oh sorry, reviewed too fast.

ronjouch commented on src/cli.test.ts:
```diff
@@ -0,0 +1,292 @@
+    const args = parseArgs(initArgs(['https://google.com', `--${arg}`, value]));
+    expect(args[arg]).toBe(value);
+
+    // Mock console.error so as not to pullute the log with the yargs help text
```
```suggestion
    // Mock console.error to not pollute the log with the yargs help text
```

ronjouch commented on src/cli.test.ts:
```diff
@@ -0,0 +1,292 @@
+import { initArgs, parseArgs } from './cli';
+import { parseJson } from './utils/parseUtils';
+
+describe('initArgs + parseArgs', () => {
```
:clap: , nice job on these tests

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+      choices: supportedArchs,
+      defaultDescription: "current node's compiled cpu architecture",
+      description: 'The cpu architecture to build for',
+      group: 'App Creation Options',
```
Two things:

1. Let's extract groups into a few consts, to avoid a typo in a group name accidentally creating a spurious extra group
2. What about adding a few visible extra "delimiter-ish" chars at the end of each group, to improve readability by making groups super visible? E.g. `group: 'App Creation Options ====='. Should not be too long, else we'll get into wrapping issues.
    1. Or (if offered by `yargs`) can groups be bolded/colored by `yargs` ?

TheCleric replied:
I think I have an easier solution for 1. Instead of specifying them in the option you can declare a group of options like:
```javascript
.group(['flash', 'flash-path'], 'Flash Options (DEPRECATED)')
```

And then future options would just need to be added to their respective group's array.

For 2, it looks like yargs supports Chalk command line coloring (if we want to `npm install chalk`, see https://github.com/yargs/yargs/issues/251), as well it looks like we could monkey with the category names to achieve what we want like so:

```
*****************
* Debug Options *
*****************
      --crash-reporter  remote server URL to send crash reports                           [string]
      --verbose         enable verbose/debug/troubleshooting logs       [boolean] [default: false]
```

ronjouch replied:
<Usual blah about limiting adding deps>, `chalk` is a total tree of 6 deps. What do you think of something dumber like my original suggestion, of using a few pre/post characters (like `=====` or `-----` or `*****`) to do the job?

TheCleric replied:
I'm mostly fine with it. To be clear, my example above is not using chalk, just pre/post chars.

ronjouch replied:
> I'm mostly fine with it. To be clear, my example above is not using chalk, just pre/post chars.

@TheCleric ooooh okay, alright then. To keep on picking nits, I see the intent of

```
*****************
* Debug Options *
*****************
```

, but it goes a bit far in the other direction, causing 3 lines per section instead of 1. I'm partial to either 

```
Debug Options =====
```

or

```
Debug Options
=============
```

TheCleric replied:
> ```
> ===== Debug Options =====
> ```

I like this one the best! 😄

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+      alias: 'arch',
+      choices: supportedArchs,
+      defaultDescription: "current node's compiled cpu architecture",
+      description: 'The cpu architecture to build for',
```
```suggestion
      description: 'The CPU architecture to build for',
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    .option('a', {
+      alias: 'arch',
+      choices: supportedArchs,
+      defaultDescription: "current node's compiled cpu architecture",
```
Conciseness nit:

```suggestion
      defaultDescription: "current Node's arch",
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    )
+    .example(
+      '$0 <targetUrl> -p <platform> -a <arch>',
+      'Make an app from <targetUrl> for the os <platform> and cpu architecture <arch>',
```
```suggestion
      'Make an app from <targetUrl> for the OS <platform> and CPU architecture <arch>',
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    .option('i', {
+      alias: 'icon',
+      description:
+        'the icon file to use as the icon for the app (should be a .png, on macOS can also be an .icns)',
```
```suggestion
        'the icon file to use as the icon for the app (.ico on Windows, .icns/.png on macOS, .png on Linux)',
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+      type: 'string',
+    })
+    .option('bookmarks-menu', {
+      description: 'Path to JSON configuration file for the bookmarks menu',
```
```suggestion
      description: 'Path to JSON config file for a bookmarks menu. See https://github.com/nativefier/nativefier/blob/master/docs/api.md#bookmarks-menu',
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    })
+    .option('max-height', {
+      defaultDescription: 'unlimited',
+      description: 'set window maximum height; in pixels',
```
Nit: consistency with the other options ending with "in pixels"

```suggestion
      description: 'set window maximum height in pixels',
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    })
+    .option('single-instance', {
+      default: false,
+      description: 'allow only a single instance of the application',
```
Conciseness nit:

```suggestion
      description: 'allow only a single instance of the app',
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    .option('file-download-options', {
+      coerce: parseJson,
+      description:
+        'a JSON string of key/value pairs to be set as file download options. See https://github.com/sindresorhus/electron-dl for available options.',
```
Conciseness nit:

```suggestion
        'a JSON string defining file download options. See https://github.com/sindresorhus/electron-dl',
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    .option('clear-cache', {
+      default: false,
+      description:
+        'prevent the application from preserving cache between launches',
```
Conciseness nit:
```suggestion
        'prevent the app from preserving cache between launches',
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    .option('disable-old-build-warning-yesiknowitisinsecure', {
+      default: false,
+      description:
+        'Disables warning when opening an app made with an old version of Nativefier. Nativefier uses the Chrome browser (through Electron), and it is dangerous to keep using an old version of it.)',
```
```suggestion
        'disables warning when opening an app made too long ago. Nativefier uses the Chrome browser (through Electron), and it is dangerous to keep using an old version of it',
```

TheCleric replied:
> Here and everywhere: please make a pass of homogenizing:
> 
> 1. The tense of descriptions. Currently it is a mix of infinitive ("ignore", "make"), and singular 3rd person ("disables")
> 2. Capitalization of `description`'s first character. Currently, some descriptions start capitalized, some don't.
> 
> I don't have a preference; up to you to see what fits best.

No problem. In this case I was trying to keep the existing verbiage, but I'll look at cleaning it up.

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    .option('disable-old-build-warning-yesiknowitisinsecure', {
+      default: false,
+      description:
+        'Disables warning when opening an app made with an old version of Nativefier. Nativefier uses the Chrome browser (through Electron), and it is dangerous to keep using an old version of it.)',
```
Here and everywhere: please make a pass of homogenizing:

1. The tense of descriptions. Currently it is a mix of infinitive ("ignore", "make"), and singular 3rd person ("disables")
2. Capitalization of `description`'s first character. Currently, some descriptions start capitalized, some don't.

I don't have a preference; up to you to see what fits best.

TheCleric replied:
> Here and everywhere: please make a pass of homogenizing:
> 
> 1. The tense of descriptions. Currently it is a mix of infinitive ("ignore", "make"), and singular 3rd person ("disables")
> 2. Capitalization of `description`'s first character. Currently, some descriptions start capitalized, some don't.
> 
> I don't have a preference; up to you to see what fits best.

No problem. In this case I was trying to keep the existing verbiage, but I'll look at cleaning it up.

ronjouch replied:
> In this case I was trying to keep the existing verbiage, but I'll look at cleaning it up.

Yup yup, I know this is all existing code. But it's a nice time to clean it up 🙂.

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+      group: '(In)Security Options',
+      type: 'boolean',
+    })
+    // Flash Options (DEPRECATED)
+    .option('flash', {
+      default: false,
+      deprecated: true,
+      description: 'enables Adobe Flash',
+      group: 'Flash Options (DEPRECATED)',
+      type: 'boolean',
+    })
+    .option('flash-path', {
+      deprecated: true,
+      description: 'path to Chrome flash plugin; find it in `chrome://plugins`',
+      group: 'Flash Options (DEPRECATED)',
+      normalize: true,
+      type: 'string',
+    })
```
If `yargs` has a feature for this, can we actually hide these options (but keep honoring them if passed, by longtime users, or people reading API.md) ?

They're deprecated, after all, and consuming attention here.

TheCleric replied:
Yep! You can mark them as `hidden: true`.

I'll do that for these deprecated ones if you'd like.

ronjouch replied:
> Yep! You can mark them as `hidden: true`.
> 
> I'll do that for these deprecated ones if you'd like.

Great. Does that make sense, or would you rather keep them?

TheCleric replied:
Nope I think hiding them and marking them as deprecated in API.md is the way to go.

ronjouch replied:
> If `yargs` has a feature for this, can we actually hide these options (but keep honoring them if passed, by longtime users, or people reading API.md) ?

@TheCleric by the way, I laid out in [HACKING.md](https://github.com/nativefier/nativefier/blob/master/HACKING.md) the respect for backwards compatibility (said differently, limiting breaking changes) I'd like contributors to have. It was unsaid and only communicated implicitly until now, from now on it's written:

> Do your best to **limit breaking changes**. Only introduce breaking changes when necessary, when required by deps, or when not breaking would be unreasonable. When you can, support the old thing forever. For example, keep maintaining old flags; to "replace" an flag you want to replace with a better version, you should keep honoring the old flag, and massage it to pass parameters to the new flag, maybe using a wrapper/adapter.
>
> Yes, our code will get a tiny bit uglier than it could have been with a hard breaking change, but that would be to ignore our users. Introducing breaking changes willy nilly is a comfort to us developers, but is disrespectful to end users who must constantly bend to the flow of breaking changes pushed by _all their software_ who think it's "just one breaking change".

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
-// package.json is `require`d to let tsc strip the `src` folder by determining
-// baseUrl=src. A static import would prevent that and cause an ugly extra "src" folder
-const packageJson = require('../package.json'); // eslint-disable-line @typescript-eslint/no-var-requires
+  // Have to access argv to get yargs to actually process the args passed to it
```
```suggestion
  // We must access argv in order to get yargs to actually process args
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
-}
+export function parseArgs(args: yargs.Argv<any>): any {
+  const parsed = { ...args.argv };
+  parsed.targetUrl = parsed._.length > 0 ? parsed._[0].toString() : '';
+  parsed.out = parsed._.length > 1 ? parsed._[1] : '';
```
What is `parsed._` in yargsland? A comment / link to docs will be welcome.

TheCleric replied:
Will do, though for reference it's how it handles positional arguments (such as targetUrl and outputDirectory)

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+  if (parsed.upgrade && parsed.targetUrl !== '') {
+    let targetAndUpgrade = false;
+    if (parsed.out === '') {
+      // If we're upgrading the first positional args might be the outputDirectory, so swap these if we can
```
```suggestion
      // If we're upgrading, the first positional args might be the outputDirectory, so swap these if we can
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+      type: 'string',
+    })
+    .option('u', {
+      alias: 'user-agent',
```
I was seeing

```
Internal Browser Options
      --file-download-options  a JSON string of key/value pairs to be set as file download options. See
                               https://github.com/sindresorhus/electron-dl for available options.       [string]
      --honest                 prevent the normal changing of the user agent string to appear as a regular
                               Chrome browser                                         [boolean] [default: false]
      --inject                 path to a CSS/JS file to be injected. Pass multiple times to inject multiple
                               files.                                                                    [array]
      --lang                   set the language or locale to render the web site as (e.g., "fr", "en-US", "es",
                               etc.)                       [string] [default: os language at runtime of the app]
  -u, --user-agent             set the app user agent string                                            [string]
```

And it made me think: would yargs support aliasing options? It'd be nice to alias `--user-agent-honest` with `--honest` (with `--user-agent-honest` being the "main" one). Goals:

1. Proximity of `--user-agent` and `--user-agent-honest` (right now, because of alphabetical sorting --which is good--, there are three options between them, hindering discoverability 😕)
2. Backwards compatibility (we want to keep honoring `--honest` for users using it, but we want to direct new users to `--user-agent-honest`

TheCleric replied:
> I was seeing
> 
> ```
> Internal Browser Options
>       --file-download-options  a JSON string of key/value pairs to be set as file download options. See
>                                https://github.com/sindresorhus/electron-dl for available options.       [string]
>       --honest                 prevent the normal changing of the user agent string to appear as a regular
>                                Chrome browser                                         [boolean] [default: false]
>       --inject                 path to a CSS/JS file to be injected. Pass multiple times to inject multiple
>                                files.                                                                    [array]
>       --lang                   set the language or locale to render the web site as (e.g., "fr", "en-US", "es",
>                                etc.)                       [string] [default: os language at runtime of the app]
>   -u, --user-agent             set the app user agent string                                            [string]
> ```
> 
> And it made me think: would yargs support aliasing options? It'd be nice to alias `--user-agent-honest` with `--honest` (with `--user-agent-honest` being the "main" one). Goals:
> 
> 1. Proximity of `--user-agent` and `--user-agent-honest` (right now, because of alphabetical sorting --which is good--, there are three options between them, hindering discoverability 😕)
> 2. Backwards compatibility (we want to keep honoring `--honest` for users using it, but we want to direct new users to `--user-agent-honest`

So like this?
```
Internal Browser Options
      --file-download-options        a JSON string of key/value pairs to be set as file download options. See https://github.com/sindresorhus/electron-dl for available options.
                                                                                                                                                                          [string]
      --inject                       path to a CSS/JS file to be injected. Pass multiple times to inject multiple files.                                                   [array]
      --lang                         set the language or locale to render the web site as (e.g., "fr", "en-US", "es", etc.)  [string] [default: os language at runtime of the app]
  -u, --user-agent                   set the app user agent string                                                                                                        [string]
      --user-agent-honest, --honest  prevent the normal changing of the user agent string to appear as a regular Chrome browser                         [boolean] [default: false]
```

ronjouch replied:
> So like this?

I had in mind not keeping `--honest` (but still honoring it), but if what you show is what's easy with `yargs`, perfect.

TheCleric replied:
We could do it with a bit of extra code (make a separate option for `--honest` that is hidden, and when set, manually pass it on to `--user-agent-honest`). Up to you.

ronjouch replied:
> We could do it with a bit of extra code (make a separate option for `--honest` that is hidden, and when set, manually pass it on to `--user-agent-honest`). Up to you.

@TheCleric nah, let's not go here. What you suggested is great and actually better to limit confusion for existing users.

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    .option('app-copyright', {
+      description:
+        '(macOS, windows only) a human-readable copyright line for the app. Maps to `LegalCopyright` metadata property on Windows, and `NSHumanReadableCopyright` on macOS',
+      group: 'Platform Specific Options',
```
Here and everywhere (or only once if you agree moving this to a const):

```suggestion
      group: 'Platform-Specific Options',
```

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    .example(
+      '$0 <targetUrl> -p <platform> -a <arch>',
+      'Make an app from <targetUrl> for the os <platform> and cpu architecture <arch>',
+    )
```
Let's add a link here to `CATALOG.md`! 🙂

```
    .example(
      'for more examples and help...',
      'see https://github.com/nativefier/nativefier/blob/master/CATALOG.md',
    )

ronjouch commented on src/cli.ts:
```diff
@@ -1,51 +1,537 @@
+    })
+    .version()
+    .help()
+    .group(['version', 'help'], 'Other Options')
```
Nit, maybe internal to yargs and not a big deal if not polishable: why do these show up as `[boolean]` in the final result?

```
Other Options
      --version  Show version number                 [boolean]
      --help     Show help                           [boolean]
                                                THIS ^^^^^^^^^
```

TheCleric replied:
Because they technically are, as in is it true you want to show the help/version?

I'll see what I can do about dropping it, but I think that's the only way to have a flag that doesn't take a value. (i.e., `--help` not `--help <value>`

ronjouch replied:
Alright. But again, it's a super nit and not a big deal. If this is what it is, no problem.

ronjouch reviewed (changes requested):
Looks pretty cool 👍. A couple things & questions, mostly nits.

TheCleric commented:
Okay, I think I either addressed your comments, or responded to them. Let me know what you're thinking.

ronjouch commented:
> Okay, I think I either addressed your comments, or responded to them. Let me know what you're thinking.

@TheCleric 👍, I have a few replies.

TheCleric commented:
> > Okay, I think I either addressed your comments, or responded to them. Let me know what you're thinking.
> 
> @TheCleric 👍, I have a few replies.

Are we all good now?

ronjouch commented on src/cli.ts:
```diff
@@ -488,6 +521,10 @@ export function initArgs(argv: string[]): yargs.Argv<any> {
 
+function decorateYargOptionGroup(value: string): string {
+  return `====== ${value} ======`;
+}
```
👍

ronjouch reviewed (approved):
Wheeeeeeeeeeeeee
